sound/soc/sof/ipc4-topology.c

```diff
@@ -179,57 +181,71 @@ static int sof_ipc4_get_audio_fmt(struct snd_soc_component *scomp,
 				  struct sof_ipc4_base_module_cfg *module_base_cfg)
 {
 	struct sof_ipc4_pin_format *out_format;
-	int audio_fmt_num = 0;
 	int ret;
 
-	ret = sof_update_ipc_object(scomp, &audio_fmt_num,
+	ret = sof_update_ipc_object(scomp, available_fmt,
 				    SOF_AUDIO_FMT_NUM_TOKENS, swidget->tuples,
-				    swidget->num_tuples, sizeof(audio_fmt_num), 1);
-	if (ret || audio_fmt_num <= 0) {
-		dev_err(scomp->dev, "Invalid number of audio formats: %d\n", audio_fmt_num);
-		return -EINVAL;
+				    swidget->num_tuples, sizeof(available_fmt), 1);
+	if (ret) {
+		dev_err(scomp->dev, "Failed to parse audio format token count\n");
+		return ret;
 	}
-	available_fmt->audio_fmt_num = audio_fmt_num;
 
-	dev_dbg(scomp->dev, "Number of audio formats: %d\n", available_fmt->audio_fmt_num);
 
-	available_fmt->sink_audio_fmts = kcalloc(available_fmt->audio_fmt_num,
-						 sizeof(struct sof_ipc4_audio_format), GFP_KERNEL);
-	if (!available_fmt->sink_audio_fmts)
-		return -ENOMEM;
-
-	ret = sof_update_ipc_object(scomp, available_fmt->sink_audio_fmts,
-				    SOF_IN_AUDIO_FORMAT_TOKENS, swidget->tuples,
-				    swidget->num_tuples, sizeof(struct sof_ipc4_pin_format),
-				    available_fmt->audio_fmt_num);
-	if (ret) {
-		dev_err(scomp->dev, "parse base_config audio_fmt tokens failed %d\n", ret);
-		goto err_in;
+	if (!available_fmt->num_sink_formats && !available_fmt->num_source_formats) {
+		dev_err(scomp->dev, "No sink and source pin formats set in topology\n");
+		return -EINVAL;
 	}
 
-	dev_dbg(scomp->dev, "Get input audio formats for %s\n", swidget->widget->name);
-	sof_ipc4_dbg_audio_format(scomp->dev, available_fmt->sink_audio_fmts,
-				  available_fmt->audio_fmt_num);
+	dev_dbg(scomp->dev,
+		"Number of sink audio formats: %d. Number of source audio formats: %d\n",
+		available_fmt->num_sink_formats, available_fmt->num_source_formats);
 
-	out_format = kcalloc(available_fmt->audio_fmt_num, sizeof(*out_format), GFP_KERNEL);
-	if (!out_format) {
-		ret = -ENOMEM;
-		goto err_in;
+	if (available_fmt->num_sink_formats) {
+		available_fmt->sink_audio_fmts = kcalloc(available_fmt->num_sink_formats,
+							 sizeof(struct sof_ipc4_pin_format),
+							 GFP_KERNEL);
+		if (!available_fmt->sink_audio_fmts)
+			return -ENOMEM;
+
+		ret = sof_update_ipc_object(scomp, available_fmt->sink_audio_fmts,
+					    SOF_IN_AUDIO_FORMAT_TOKENS, swidget->tuples,
+					    swidget->num_tuples,
+					    sizeof(struct sof_ipc4_pin_format),
+					    available_fmt->num_sink_formats);
+		if (ret) {
+			dev_err(scomp->dev, "parse sink audio_fmt tokens failed %d\n", ret);
+			goto err_in;
+		}
+
+		dev_dbg(scomp->dev, "Get input audio formats for %s\n", swidget->widget->name);
```

Comment:
s/Get/Got

Reply:
It is likely me again, but I rarely associate `sink` with `input` as the same thing.